### PR TITLE
test commodity interfaces

### PR DESCRIPTION
**Pull request by bereba**
Added script and input data to check commodity interfaces between model structure of SEDOS sectors.
Please check the script and the location of its input data, which basically are the process_set from the model structure excel.

### CODE REVIEW

**Comment by FelixMau** [comment]
I updated your code to use the Structures class so we can add the functionality to the adapter directly. I hope you don't mind @bereba

File: data_adapter/structure.py [hunk 1]
```diff
@@ -150,3 +152,149 @@ def _init_parameters(self, parameter_sheet: str) -> dict:
...
+            Patch(facecolor="green", edgecolor="black", label="Output"),
+            Patch(facecolor="yellow", edgecolor="black", label="Input & Output"),
+        ]
+        fig.legend(handles=legend_elements, loc="upper left", fontsize=12)
+        plt.tight_layout()
+
+        # Save the plot as svg
+        plt.show()
+
+    def get_commodity_diff(self, input_processes: list = ["source", "import"], output_processes: list = ["sink"]):
```
Comment (henhuy) [inline data_adapter/structure.py]: One question: Do we need to set input and output processes explicitly? I would think, that processes which either have only inputs or only outputs are "creating" or "destroying" commodities implicitly - that would it make easier in case some processes are not named sink/source or import...

File: tests/test_commodity_interfaces.py
Comment (henhuy) [inline tests/test_commodity_interfaces.py:1]: This is not a real test.
Better option would be to add a minimal structure with predefined missing sinks or sources and write a test which checks if missing parts are found. If you don't have the time to do so, I will merge nevertheless as this function is optional.

**Review by henhuy — DISMISSED** [review]
Only checked if its working - did not reviewed code itself.
Have two recommendations to think about.
But you can merge anyway if you want!

**Comment by henhuy** [comment]
Made matplotlib optional and moved structure file to `test_data/test_structures`

**Comment by henhuy** [comment]
Also linter is not happy with `structure.py` due to your changes...

File: data_adapter/structure.py (hunk 1 above)
Comment (FelixMau) [reply data_adapter/structure.py]: Haven't thought it that way, thank you!

**Comment by FelixMau** [comment]
Separated parts of the plot function outside the class to make the linter happy.

File: tests/test_commodity_interfaces.py [hunk 2]
```diff
@@ -2,100 +2,18 @@
...
 
-# Fill empty cells to facilitate logic - it is not considered as it has not one of the predefined category codes
-df.fillna('dummytext', inplace=True)
+structure.plot_commodity_interfaces()
 
-# Filter out processes which impair the logic
-df = df[~df['process'].str.contains('storage|transport|import|export')]
-
-# Delete square brackets from dataframe
-df['input'] = df['input'].str.replace(r'[][]', '', regex=True)
```
Comment (bereba) [inline tests/test_commodity_interfaces.py]: In my last review I additionally deleted blank spaces to avoid unintended double naming and the _orig ending for commodities to simplify the graph.

File: tests/test_commodity_interfaces.py [hunk 3]
```diff
@@ -2,100 +2,18 @@
...
+    parameter_sheet="Parameter_O1",
+    helper_sheet="Helper_O1",
+)
 
-# Fill empty cells to facilitate logic - it is not considered as it has not one of the predefined category codes
-df.fillna('dummytext', inplace=True)
+structure.plot_commodity_interfaces()
 
-# Filter out processes which impair the logic
-df = df[~df['process'].str.contains('storage|transport|import|export')]
```
Comment (bereba) [inline tests/test_commodity_interfaces.py]: In my last interface check I also filtered the following key words (due to changes in the model structure list):
storage|export|helper_source|helper_sink'

**Review by bereba — COMMENTED** [review]
I recently changed some details for the interface check in the code locally.
You might want to apply this in your new functions as well.

File: tests/test_commodity_interfaces.py (hunk 2 above)
Comment (bereba) [reply tests/test_commodity_interfaces.py]: df['input'] = df['input'].str.replace(r' ', '', regex=True)
df['output'] = df['output'].str.replace(r' ', '', regex=True)
df['input'] = df['input'].str.replace(r'_orig', '', regex=True)
df['output'] = df['output'].str.replace(r'_orig', '', regex=True)

File: data_adapter/structure.py (hunk 1 above)
Comment (bereba) [reply data_adapter/structure.py]: Maybe it would also make sense to have an additional color (e.g. blue) for commodities that only come from  imports/sources. I think for the interface check this might be an interesting value.

sinks and exports could be filtered as mentioned above. However, sources and imports should be checked. I accordingly adjust my proposed filter in the review below.

File: tests/test_commodity_interfaces.py (hunk 3 above)
Comment (bereba) [reply tests/test_commodity_interfaces.py]: As elaborated in henhuys comment my suggestion for the filters is as follows:

'storage|export||helper_sink|delivery'

As "_orig" could neglected I would also add the delivery processes.
What do you think?

File: tests/test_commodity_interfaces.py (hunk 2 above)
Comment (FelixMau) [reply tests/test_commodity_interfaces.py]: I think it might be even necessary to see the blank spaces. Since I am using the data_adapter itself we have to have a consistent naming here and might miss an opportunity to identify potential typos.

File: data_adapter/structure.py (hunk 1 above)
Comment (FelixMau) [reply data_adapter/structure.py]: I have added blue for commodities that are only outputs of processes (but don't go anywhere) and purple for commodities that have no source to them (picture below) 
Figure_1

Comment (FelixMau) [reply data_adapter/structure.py]: This function is also separated from the function for drawing the commodities. 
It is meant only to get a list of "open" commodities, therefore I would like to include sources and sinks (in opposite of my thinking above) to see what commodities are either not destroyed or not created yet.

File: tests/test_commodity_interfaces.py (hunk 3 above)
Comment (FelixMau) [reply tests/test_commodity_interfaces.py]: As I am not very deeply involved into discussions about naming and different edge cases, I don't really understand what you want to achieve with the filtering and what effect the exact process naming has. 
I don't understand what benefit comes not including sinks and sources in the plot?

I do understand that excluding `storage` makes sense to make the plot smaller. Even though it can have multiple outputs/inputs as  `helper_co2_storage_permanent` that has an output going into "nowhere" and for modelling I eventually need to be aware of such issues to help debugging.

File: data_adapter/structure.py (hunk 1 above)
Comment (bereba) [reply data_adapter/structure.py]: I am not sure if this really helps, as now all these open commodities are blue and we do not know if it is an input or output, right? We could see those processes before, if one line did only have green or red cell.
The idea of the additional color was more to highlight commodities which have no other upstream processes except the imports/sources.

If you want it more clear that there is something "open" you could maybe use another formating. Like highighting the text or something?

File: tests/test_commodity_interfaces.py (hunk 2 above)
Comment (bereba) [reply tests/test_commodity_interfaces.py]: I don't understand this point. Could you elaborate this more please?
We are talking about the naming of inputs and outputs, which might be separated with commas, right?

File: tests/test_commodity_interfaces.py (hunk 3 above)
Comment (bereba) [reply tests/test_commodity_interfaces.py]: The idea of this plot first was mainly to check if there are defined inputs which are not being created or outputs which are not being used.

Therefore, temporal (storages) or spatial (delivery) shifts are disturbing this logic, as it might appear that e.g. electricity is present in our system as input and output, although there is no process to generate it.

For the sinks and exports, it is a modeling-specific decision I guess, and if it helps you, you can also keep them.

File: tests/test_commodity_interfaces.py (hunk 2 above)
Comment (FelixMau) [reply tests/test_commodity_interfaces.py]: Yes but the whitespace trimming already happens within the data_adapter beforehand, as the plot function is now part of the class I thought it is better to reuse already existing parts of code 
To improve consistency I want to use exactly what is going to be input for the data_adapter_oemof.

Comment (bereba) [reply tests/test_commodity_interfaces.py]: Ah okay now I understand. Makes sense.

File: tests/test_commodity_interfaces.py (hunk 3 above)
Comment (FelixMau) [reply tests/test_commodity_interfaces.py]: Ah yes now I see, thank you!

So delivery should only be a kind of transport? As in the Model structure unfortunately there are processes that seem to be also doing Conversion. Such as `helper_co2_delivery`

File: data_adapter/structure.py (hunk 1 above)
Comment (FelixMau) [reply data_adapter/structure.py]: Yes we did know what inputs/outputs were but I did not understand correctly what you wanted :)
So here is the improved version for the Steel Industry Part O1: 
Figure_1

File: tests/test_commodity_interfaces.py (hunk 3 above)
Comment (bereba) [reply tests/test_commodity_interfaces.py]: Yes, as we have a one-node model approach, the delivery is always modeled with a conversion between commodities.
For helper_co2_delivery different commodities are additionally merged.

**Comment by henhuy** [comment]
Is this ready to merge?

**Comment by FelixMau** [comment]
From my side, yes
